text/0000-synthetic-shadow-migrate.md

```diff
@@ -0,0 +1,153 @@
+
+However, if any descendant components use `'native'` mode, then `'migrate'` ceases propagating. If a component is already opting in to `'native'` mode, then it is assumed that it doesn't need any global styles for compatibility.
+
+Like `'native'`, component authors can also use `'reset'` explicitly to override a superclass's `'migrate'` value. (This is simply a product of how JavaScript classes work.)
+
+### Limitations
+
+Copying styles from the `<head>` into the shadow root does not cover all possible CSS selectors. Most notably, descendant
+selectors that cross shadow boundaries – e.g. `.foo .bar`, where `.foo` is in one shadow root and `.bar` is in another – will
+not be supported.
+
+However, this technique was already tested on a non-trivial internal Salesforce app, and there was only one component with incorrect styles. Therefore, "migrate" mode can be thought of as a "best effort" to emulate the behavior of synthetic shadow DOM, which will probably get component authors 99.9% of the way there, but without the effort required to _fully_ migrate to native shadow.
+
+Also, it is understood that copying styles into shadow roots may lead to a performance cost. However, browsers already have
+[optimizations for repeated stylesheets in shadow roots](https://github.com/whatwg/dom/issues/831#issuecomment-585489960), so this cost is not expected to be large.
+
+Finally, `'migrate'` mode does not attempt to emulate non-stylesheet differences between native and synthetic shadow. For example, it does not attempt to emulate "lazy" slots, or the nonstandard behavior of `innerHTML`, or other quirks. These are all things that `'migrate'` _could_ do in a subsequent version, possibly gated by [component-level API versioning](https://lwc.dev/guide/versioning#component-level-api-versioning).
+
+For a v1 of `'migrate'` mode, however, merely cloning stylesheets should provide the most bang for the buck. It is not a complicated feature to implement, and it is likely to cover the biggest hurdle that component authors would hit when migrating from synthetic to native shadow.
+
+### Server-side rendering
+
+In SSR mode, no attempt is made to inject particular stylesheets into the DOM at server-side render time.
+
+This means that there may be a [flash of unstyled content](https://en.wikipedia.org/wiki/Flash_of_unstyled_content) when the components
+hydrate on the client side. Also, LWC's hydration logic will have to explicitly ignore the missing `<link>`/`<style>` elements and not report
+them as hydration warnings.
+
+The downside is, of course, an ugly load experience and potential hit to [Cumulative Layout Shift](https://web.dev/articles/cls). The solution
+for component authors who are concerned about this is to move fully from `'migrate'` to `'native'` shadow mode.
+
+While it would be technically possible to have a handoff between the client and server where both are aware of which `<link>` and `<style>`

```

Comment:
Can you expand on what makes this complex?

Why can't a call to renderComponent() include passing the list of styles to clone into components?

Reply:
`renderComponent` is currently not aware of any global styles on the page. Also, there are a number of sources for global styles:

1. Stylesheets owned by the page container
2. Aura components
3. Styles rendered in another framework (e.g. Vue/Svelte/React components appending `<style>`s to the `<head>` as they render client-side (similar to Aura).

The first category would be the easiest to cover, but I'm not sure it's worth it. For a v1 anyway, it's much simpler to say "this is CSR-only."

Reply:
FOUC (or mistyled?) is the thing I'd like to avoid, if possible. 

In the case of SSR, is it unfair to put the onus on the invoker of `renderComponent` to provide the right list of style/link tags?

Reply:
This is totally possible to do, but I would prefer to avoid it for a v1. If it's just SLDS it's pretty easy, but some LWR sites do have custom CSS added via static resources. So to make it work, there would be a lot of moving parts.

Right now my main goal is to validate this approach for the simple case (CSR) before moving on to the complex case (SSR+hydration).